source/fab/errors.py

```diff
@@ -0,0 +1,344 @@
+
+    Error where an invalid setting, e.g. MPI, has been requested for a
+    particular tool.
+
+    :param setting_type: name of the invalid setting
+    :param tool: the tool to which setting applies
+    :param additional: optional additional information
+    """
+
+    # FIXME: improve these here and in tool_repository
+
+    def __init__(
+        self,
+        setting_type: str,
+        tool: Union["Category", "Tool", str],
+        additional: Optional[str] = None,
+    ) -> None:
+        self.setting_type = setting_type
+
+        message = f"invalid {setting_type}"
+        if additional:
+            message += f" {additional}"
+
+        super().__init__(tool, message)
+
+
+class FabUnknownLibraryError(FabError):
+    """An unknown library has been requested.
+
+    Error where an library which is not known to the current Linker
+    instance is requested.
+
+    :param library: the name of the unknown library
+    """
+
+    def __init__(self, library: str) -> None:
+        self.library = library
+        super().__init__(f"unknown library {library}")
+
+
+class FabCommandError(FabError):
+    """An error was encountered running a subcommand.
+
+    Error where a subcommand run by the fab framework returned with a
+    non-zero exit code.  The exit code plus any data from the stdout
+    and stderr streams are retained to allow them to be passed back up
+    the calling stack.
+
+    :param command: the command being run
+    :param code: the command return code from the OS
+    :param output: output stream from the command
+    :param error: error stream from the command
+    """
+
+    def __init__(
+        self,
+        command: str,
+        code: int,
+        output: Union[str, bytes, None],
+        error: Union[str, bytes, None],
+        cwd: Optional[Union[str, Path]] = None,
+    ) -> None:
+        if isinstance(command, list):
+            self.command: str = " ".join(command)
+        else:
+            self.command = str(command)
+        self.code = int(code)
+        self.output = self._decode(output)
+        self.error = self._decode(error)
+        self.cwd = cwd
+        super().__init__(f"command {repr(self.command)} returned {code}")
+
+    def _decode(self, value: Union[str, bytes, None]) -> str:
+        """Convert from bytes to a string as necessary."""
+        if value is None:
+            return ""
+        if isinstance(value, bytes):
+            return value.decode()
+        return value
+
+
+class FabCommandNotFound(FabError):
+    """Target command could not be found by subprocess.
+
+    Error where the target command passed to a subprocess call could
+    not be found.
+
+    :param command: the target command.  For clarity, only the first
+        item is used in the error message but the entire command is
+        preserved for inspection by the caller.
+    """
+
+    def __init__(self, command: str) -> None:
+        self.command = command
```

Comment:
Homogenise argument to either string or list of strings. Latter would remove the need for the test to extract target.